src/app/modules/sgid/sgid.service.ts

```diff
@@ -0,0 +1,194 @@
+import { SgidClient } from '@opengovsg/sgid-client'
+import fs from 'fs'
+import { err, ok, Result, ResultAsync } from 'neverthrow'
+
+import { sgid } from '../../config/config'
+import { createLoggerWithLabel } from '../../config/logger'
+import { ApplicationError } from '../core/core.errors'
+
+import {
+  SgidCreateRedirectUrlError,
+  SgidFetchAccessTokenError,
+  SgidFetchUserInfoError,
+  SgidInvalidJwtError,
+  SgidInvalidStateError,
+  SgidVerifyJwtError,
+} from './sgid.errors'
+import { isSgidJwtPayload } from './sgid.util'
+
+const logger = createLoggerWithLabel(module)
+
+export class SgidService {
+  private client: SgidClient
+
+  private publicKey: string | Buffer
+
+  private cookieDomain: string
+  private cookieMaxAge: number
+  private cookieMaxAgePreserved: number
+
+  constructor({
+    cookieDomain,
+    cookieMaxAge,
+    cookieMaxAgePreserved,
+    privateKey,
+    publicKey,
+    ...sgidOptions
+  }: {
+    endpoint: string
+    clientId: string
+    clientSecret: string
+    privateKey: string
+    publicKey: string
+    redirectUri: string
+    cookieDomain: string
+    cookieMaxAge: number
+    cookieMaxAgePreserved: number
+  }) {
+    this.client = new SgidClient({
+      ...sgidOptions,
+      privateKey: fs.readFileSync(privateKey),
+    })
+    this.publicKey = fs.readFileSync(publicKey)
+    this.cookieDomain = cookieDomain
+    this.cookieMaxAge = cookieMaxAge
+    this.cookieMaxAgePreserved = cookieMaxAgePreserved
+  }
+
+  createRedirectUrl(state: string): Result<string, SgidCreateRedirectUrlError> {
+    const logMeta = {
+      action: 'createRedirectUrl',
+      state,
+    }
+    const result = this.client.authorizationUrl(state)
+    if (typeof result.url === 'string') {
+      return ok(result.url)
+    } else {
+      logger.error({
+        message: 'Error while creating redirect URL',
+        meta: logMeta,
+        error: result,
+      })
+      return err(new SgidCreateRedirectUrlError())
+    }
+  }
+
+  parseState(
+    state: string,
+  ): Result<{ formId: string; rememberMe: boolean }, SgidInvalidStateError> {
+    const [formId, rememberMeStr] = state.split(',')
+    const rememberMe = rememberMeStr === 'true'
+    return formId
+      ? ok({ formId, rememberMe })
+      : err(new SgidInvalidStateError())
+  }
+
+  token(
```

Comment:
nit: increase the clarity of what this function does with a more explicit name? Should it be called `retrieveAccessToken` instead?